{"cells":[{"cell_type":"markdown","id":"a86556f9","metadata":{"id":"a86556f9"},"source":["# Bulk RNA Seq Analysis\n","\n","\n","The ultimate goal of this exercise and the accompanying R script is to generate figure 5 from this paper with two replicates each from the MBC and prePB groups https://www.nature.com/articles/s41375-021-01234-0"]},{"cell_type":"markdown","source":["# Mount your Google Drive"],"metadata":{"id":"M48v9Yq3N6cl"},"id":"M48v9Yq3N6cl"},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/gdrive')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Wlisj9OHN6yL","executionInfo":{"status":"ok","timestamp":1694091749437,"user_tz":240,"elapsed":16409,"user":{"displayName":"Dev Majumdar","userId":"17609659741311115457"}},"outputId":"92c6c7c1-2cf2-4e97-ffde-d92cd46d8ae0"},"id":"Wlisj9OHN6yL","execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/gdrive\n"]}]},{"cell_type":"markdown","source":["# Create directory for the course"],"metadata":{"id":"_Xe41-vEqI2o"},"id":"_Xe41-vEqI2o"},{"cell_type":"code","source":["!mkdir /content/gdrive/MyDrive/colab\n","!mkdir /content/gdrive/MyDrive/colab/compgen"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"bE0264QbqHSW","executionInfo":{"status":"ok","timestamp":1694091751946,"user_tz":240,"elapsed":328,"user":{"displayName":"Dev Majumdar","userId":"17609659741311115457"}},"outputId":"bbbb1520-cb85-4cea-c5c8-ce07c84658e7"},"id":"bE0264QbqHSW","execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["mkdir: cannot create directory ‘/content/gdrive/MyDrive/colab’: File exists\n","mkdir: cannot create directory ‘/content/gdrive/MyDrive/colab/compgen’: File exists\n"]}]},{"cell_type":"markdown","source":["# Set up Anaconda and bioconda\n","You must run this every session, as well as subsequent installations."],"metadata":{"id":"HduGYM8eZbMD"},"id":"HduGYM8eZbMD"},{"cell_type":"code","source":["%cd /usr/local/\n","\n","!wget -c https://repo.continuum.io/archive/Anaconda3-5.1.0-Linux-x86_64.sh\n","!chmod +x Anaconda3-5.1.0-Linux-x86_64.sh\n","!bash ./Anaconda3-5.1.0-Linux-x86_64.sh -b -f -p /usr/local/\n","\n","import sys\n","\n","sys.path.append(\"/usr/local/lib/python3.6/site-packages/\")\n","\n","!conda config --add channels defaults\n","!conda config --add channels bioconda\n","!conda config --add channels conda-forge\n","%cd /content/gdrive/MyDrive/colab"],"metadata":{"id":"twdR4xpVnXgo"},"id":"twdR4xpVnXgo","execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Create and enter RNASeq directory\n"],"metadata":{"id":"utx2EdWzc0vd"},"id":"utx2EdWzc0vd"},{"cell_type":"code","source":["%cd /content/gdrive/MyDrive/colab/compgen\n","#!rm -r /content/gdrive/MyDrive/colab/compgen/RNASeq\n","!mkdir /content/gdrive/MyDrive/colab/compgen/RNASeq\n","%cd /content/gdrive/MyDrive/colab/compgen/RNASeq\n"],"metadata":{"id":"7Je5Pl8Tc1ER","executionInfo":{"status":"ok","timestamp":1694097196413,"user_tz":240,"elapsed":698,"user":{"displayName":"Dev Majumdar","userId":"17609659741311115457"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"645b4cf5-0d1a-4362-de34-c83b9e231808"},"id":"7Je5Pl8Tc1ER","execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["/content/gdrive/MyDrive/colab/compgen\n","/content/gdrive/MyDrive/colab/compgen/RNASeq\n"]}]},{"cell_type":"markdown","source":["# Install fastq-dump"],"metadata":{"id":"Li22AAzqKXb8"},"id":"Li22AAzqKXb8"},{"cell_type":"code","source":["%cd /\n","!wget https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-ubuntu64.tar.gz\n","!gunzip sratoolkit.current-ubuntu64.tar.gz\n","!tar -xvf sratoolkit.current-ubuntu64.tar"],"metadata":{"id":"cBQ6Qll1KXxV"},"id":"cBQ6Qll1KXxV","execution_count":null,"outputs":[]},{"cell_type":"markdown","id":"3a1bb7ad","metadata":{"id":"3a1bb7ad"},"source":["# Download Fastq files"]},{"cell_type":"code","source":["%%bash\n","cd /content/gdrive/MyDrive/colab/compgen/RNASeq\n","LIST=$(cat accessions.txt)\n","mkdir fastq\n","cd /\n","rm -r fasterq.tmp.9cc98f13d0f0.49085/\n","for i in ${LIST}\n","    do\n","        ./sratoolkit.3.0.7-ubuntu64/bin/fasterq-dump ${i}\n","        echo ${i}' downloaded'\n","        mv /${i}_1.fastq /content/gdrive/MyDrive/colab/compgen/RNASeq/fastq/${i}_1.fastq\n","        mv /${i}_2.fastq /content/gdrive/MyDrive/colab/compgen/RNASeq/fastq/${i}_2.fastq\n","    done"],"metadata":{"id":"Z4I7IQsVH5rc"},"id":"Z4I7IQsVH5rc","execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Install salmon"],"metadata":{"id":"qp3eEMVDvd-O"},"id":"qp3eEMVDvd-O"},{"cell_type":"markdown","source":[],"metadata":{"id":"grRgc-DixCMK"},"id":"grRgc-DixCMK"},{"cell_type":"code","source":["%%bash\n","cd /\n","conda install -y salmon"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"F4T9WhQTxCeJ","executionInfo":{"status":"ok","timestamp":1694096413682,"user_tz":240,"elapsed":131464,"user":{"displayName":"Dev Majumdar","userId":"17609659741311115457"}},"outputId":"509b6c48-027c-4f47-b6ca-23b20df437f0"},"id":"F4T9WhQTxCeJ","execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Solving environment: ...working... done\n","\n","# All requested packages already installed.\n","\n"]},{"output_type":"stream","name":"stderr","text":["\n","\n","==> WARNING: A newer version of conda exists. <==\n","  current version: 4.4.10\n","  latest version: 23.7.3\n","\n","Please update conda by running\n","\n","    $ conda update -n base conda\n","\n","\n"]}]},{"cell_type":"markdown","id":"83766ab8","metadata":{"id":"83766ab8"},"source":["# Create config.txt and samples.txt\n","This code block creates a config file and a list of samples so that most of the rest of the script can be autonomous. This code chunk is written in python.\n","\n","Config.txt will contain parameters that will be loaded into a subsequent bash script and used to complete variable names in order to download gencode files and properly format file names.\n","Samples.txt will contain a list of your sample names.\n","\n","Instructions:\n","- Edit species to reflect human or mouse.\n","- Edit version to reflect the current gencode release. Do NOT include the 'M' prefix to the mouse release. Note: you can check the current releases here https://www.gencodegenes.org/.\n","- Edit the fastqExtension. If your fastq file names have a ton of junk in the middle of them, this refers to that junk. Example below.\n","- Edit the list of your samples. Example below.\n","\n","The fastqExtension and samples list are used in the salmon quant code blocks. Check them to see more about why they are formatted the way they are.\n","\n","Example:\n","Let's say you have three paired end samples to quantify: BL, H6, and L24. Your list of fastqs looks like this:\n","- BLa739_9747_AHFD7D_1.fastq\n","- BLa739_9747_AHFD7D_2.fastq\n","- H6a739_9747_AHFD7D_1.fastq\n","- H6a739_9747_AHFD7D_2.fastq\n","- L24a739_9747_AHFD7D_1.fastq\n","- L24a739_9747_AHFD7D_2.fastq\n","\n","Your variables in the script below would then be:\n","\n","fastqExtension = 'a739_9747_AHDFD7D_'\n","\n","samples = {'BL', 'H6', 'L6']\n","\n","Note: If you have paired end reads, do NOT list each fastq separately! List each SAMPLE separately."]},{"cell_type":"code","execution_count":null,"id":"ab0d3715","metadata":{"id":"ab0d3715","outputId":"8b312b31-a4c7-4b10-b8ae-a27632217d4a","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1694108528188,"user_tz":240,"elapsed":249,"user":{"displayName":"Dev Majumdar","userId":"17609659741311115457"}}},"outputs":[{"output_type":"stream","name":"stdout","text":["/content/gdrive/MyDrive/colab/compgen/RNASeq\n","Error: you formatted your species entry incorrectly\n","Done\n"]}],"source":["%cd /content/gdrive/MyDrive/colab/compgen/RNASeq\n","species = 'humn'\n","gencodeVersion = 44\n","fastqExtension = '_CKDL230019755-1A_H75TNDSX7_L2_'\n","samples = ['sixA', 'sixB', 'tfourA', 'tfourB', 'zeroA', 'zeroB']\n","\n","if species == 'human':\n","    speciesfull = 'human'\n","    specieslow = 'h'\n","    speciesup = ''\n","    release = 38\n","elif species == 'mouse':\n","    speciesfull = 'mouse'\n","    specieslow = 'm'\n","    speciesup = 'M'\n","    release = 39\n","else:\n","    print('Error: you formatted your species entry incorrectly')\n","\n","config = open(\"config.txt\", \"w\")\n","configelements = [fastqExtension, gencodeVersion, speciesfull, specieslow, speciesup, release]\n","for line in configelements:\n","    config.write(\"%s\\n\" % line)\n","config.close()\n","\n","samps = open(\"samples.txt\", \"w\")\n","for line in samples:\n","    samps.write(\"%s\\n\" % line)\n","samps.close()\n","print('Done')"]},{"cell_type":"markdown","id":"db208529","metadata":{"id":"db208529"},"source":["# Download gencode files, generate decoys.txt + gene_map.csv, and run salmon index\n","If you did things right and I did things right, you should be able to just run this code block if you're working with mouse or human data. Good luck!\n","\n","That being said, you MUST verify that each step of this script actually worked before the subsequent steps will run correctly. If you just hit \"run\" and hope for the best, you may not catch your error until way downstream when all your files are empty.\n","\n","If you're NOT working with mouse or human data, you need to edit the wget lines to whatever will download your transcript and primary assembly files, as well as any subsequent file paths that point to the downloaded files."]},{"cell_type":"code","execution_count":null,"id":"cd9439f6","metadata":{"scrolled":true,"id":"cd9439f6","outputId":"8e203dba-75c4-44ae-dac8-07dbd99049be","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1694098685189,"user_tz":240,"elapsed":1453049,"user":{"displayName":"Dev Majumdar","userId":"17609659741311115457"}}},"outputs":[{"output_type":"stream","name":"stdout","text":["Process is terminated.\n"]}],"source":["%%bash\n","%cd /content/gdrive/MyDrive/colab/compgen/RNASeq\n","VERS=$(sed -n '2p' config.txt)\n","SPECIESFULL=$(sed -n '3p' config.txt)\n","SPECIESlow=$(sed -n '4p' config.txt)\n","SPECIESup=$(sed -n '5p' config.txt)\n","RELEASE=$(sed -n '6p' config.txt)\n","\n","#download needed gencode files\n","wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_${SPECIESFULL}/release_${SPECIESup}${VERS}/gencode.v${SPECIESup}${VERS}.transcripts.fa.gz\n","wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_${SPECIESFULL}/release_${SPECIESup}${VERS}/GRC${SPECIESlow}${RELEASE}.primary_assembly.genome.fa.gz\n","\n","#generate decoys.txt from the transcript and primary assembly files\n","grep \"^>\" <(gunzip -c GRC${SPECIESlow}${RELEASE}.primary_assembly.genome.fa.gz) | cut -d \" \" -f 1 > decoys.txt\n","sed -i.bak -e 's/>//g' decoys.txt\n","cat gencode.v${SPECIESup}${VERS}.transcripts.fa.gz GRC${SPECIESlow}${RELEASE}.primary_assembly.genome.fa.gz > gentrome.fa.gz\n","\n","#create gene_map.csv\n","gunzip gencode.v${SPECIESup}${VERS}.transcripts.fa.gz\n","echo ${SPECIESlow}\n","if [[ ${SPECIESlow} == \"h\" ]]; then\n","    echo 'human'\n","    grep -P -o 'ENST\\d{11}' gencode.v${SPECIESup}${VERS}.transcripts.fa > enst.txt\n","    grep -P -o 'ENSG\\d{11}' gencode.v${SPECIESup}${VERS}.transcripts.fa > ensg.txt\n","else\n","    echo 'mouse'\n","    grep -P -o 'ENSMUST\\d{11}' gencode.v${SPECIESup}${VERS}.transcripts.fa > enst.txt\n","    grep -P -o 'ENSMUSG\\d{11}' gencode.v${SPECIESup}${VERS}.transcripts.fa > ensg.txt\n","fi\n","paste -d ',' enst.txt ensg.txt > gene_map.csv\n","gzip gencode.v${SPECIESup}${VERS}.transcripts.fa\n","\n","#run salmon indexer\n","salmon index -t gentrome.fa.gz -d decoys.txt -p 12 -i salmon_index --gencode\n","echo 'Done'\n"]},{"cell_type":"markdown","id":"610bd849","metadata":{"id":"610bd849"},"source":["# Run Salmon quant on paired end reads\n","Run this code block if you have paired end reads.\n","\n","Validate it step by step FIRST so that you don't end up with a dozen unzipped fastqs.\n","\n","You may have to reformat the way the fastq's are named based on how your fastq names are formatted. Use echo statements to debug."]},{"cell_type":"code","execution_count":null,"id":"a72e174a","metadata":{"id":"a72e174a"},"outputs":[],"source":["%%bash\n","conda activate salmon\n","LIST=$(cat samples.txt)\n","FQEXT=$(sed -n '1p' config.txt)\n","for i in ${LIST}\n","    do\n","        gunzip fastq/${i}${FQEXT}1.fq.gz\n","        gunzip fastq/${i}${FQEXT}2.fq.gz\n","        salmon quant -i salmon_index -l A -1 fastq/${i}${FQEXT}1.fq -2 fastq/${i}${FQEXT}2.fq --validateMappings -o ${i}\n","        gzip fastq/${i}${FQEXT}1.fq\n","        gzip fastq/${i}${FQEXT}2.fq\n","    done"]},{"cell_type":"markdown","id":"1e6c20af","metadata":{"id":"1e6c20af"},"source":["# Run salmon quant on single end reads\n","Run this code block if you have single end reads.\n","\n","Validate it step by step FIRST so that you don't end up with a dozen unzipped fastqs.\n","\n","You may have to reformat the way the fastq's are named based on how your fastq names are formatted. Use echo statements to debug."]},{"cell_type":"code","execution_count":null,"id":"0994a225","metadata":{"id":"0994a225","outputId":"70756e3c-1a9b-4f4c-9d17-5eccd212b360"},"outputs":[{"name":"stdout","output_type":"stream","text":["fastq/sixA_CKDL230019755-1A_H75TNDSX7_L2_2.fq\n","Process is interrupted.\n"]}],"source":["%%bash\n","conda activate salmon\n","LIST=$(cat samples.txt)\n","FQEXT=$(sed -n '1p' config.txt)\n","for i in ${LIST}\n","    do\n","        gunzip fastq/${i}${FQEXT}.fq.gz\n","        salmon quant -i salmon_index -l A -r fastq/${i}${FQEXT}.fq --validateMappings -o ${i}\n","        gzip fastq/${i}${FQEXT}.fq\n","    done"]},{"cell_type":"markdown","id":"03c8c84a","metadata":{"id":"03c8c84a"},"source":["# Generate SRARunTable.txt files\n","DeSeq2 will eventually rearrange your data so that it is alphabetical and then treats whatever comes first as the control. Make sure your controls are listed FIRST and are the first of your samples alphabetically. You may need to edit the names of your salmon outputs to reflect this. Edit these lists in the code block below according to the following:\n","\n","- samples: This should be a list of your samples. Unless you need to make changes for alphabetization, this should match your sample list from the first code block.\n","- treatments: Every value is either \"control\" or \"treatment\" in this list. List positions correspond to sample list positions.\n","- conditions: Each value is an experiment-specific treatment label. These are used to group samples by condition.\n","- cell_line: Each value is the cell line that the experiment sample was carried out in.\n","- files: This should be your ONLY list of a different length that does NOT correspond to positions in the other lists. This should just be a list of conditions that exist in your experiment, excluding \"control.\" However you format this should match what you tell R in subsequent analysis.\n","\n","For an example, see the start to the code block below."]},{"cell_type":"code","execution_count":null,"id":"44602278","metadata":{"id":"44602278","outputId":"e938b5f3-a765-426b-8274-75032b4a83e1"},"outputs":[{"name":"stdout","output_type":"stream","text":["Done\n"]}],"source":["samples = ['controlA','controlB','NonionA','NonionB','PBSA','PBSB','SM102eLNPA','SM102eLNPB','SM102RNAA','SM102RNAB','TCLeLNPA','TCLeLNPB','TCLRNAA','TCLRNAB']\n","treatments = [\"control\", \"control\", \"treatment\", \"treatment\", \"treatment\", \"treatment\", \"treatment\", \"treatment\",\"treatment\",\"treatment\",\"treatment\",\"treatment\",\"treatment\",\"treatment\"]\n","files = [\"nonion\",\"PBS\",\"SM102eLNP\",\"SM102RNA\",\"TCLeLNP\",\"TCLRNA\"]\n","conditions = [\"control\", \"control\",\"nonion\",\"nonion\",\"PBS\",\"PBS\",\"SM102eLNP\",\"SM102eLNP\",\"SM102RNA\",\"SM102RNA\",\"TCLeLNP\",\"TCLeLNP\",\"TCLRNA\",\"TCLRNA\"]\n","cell_line = ['mouse','mouse','mouse','mouse','mouse','mouse','mouse','mouse','mouse','mouse','mouse','mouse','mouse','mouse']\n","\n","\n","SRT = open(\"SraRunTableall.txt\", \"w\")\n","SRT.write(\"%s\\n\" % \"sample name, treatment, cell_line\")\n","for i in range(len(samples)):\n","    SRT.write(samples[i])\n","    SRT.write(', ')\n","    SRT.write(treatments[i])\n","    SRT.write(', ')\n","    SRT.write(\"%s\\n\" % cell_line[i])\n","SRT.close()\n","\n","for j in range(len(files)):\n","    filename = \"SraRunTable\"+files[j]+\".txt\"\n","    SRT = open(filename, \"w\")\n","    SRT.write(\"%s\\n\" % \"sample name, treatment, cell_line\")\n","    for i in range(len(samples)):\n","        if conditions[i] == 'control' or conditions[i] == files[j]:\n","            SRT.write(samples[i])\n","            SRT.write(', ')\n","            SRT.write(treatments[i])\n","            SRT.write(', ')\n","            SRT.write(\"%s\\n\" % cell_line[i])\n","    SRT.close()\n","\n","print(\"Done\")"]},{"cell_type":"code","execution_count":null,"id":"a1c9d473","metadata":{"id":"a1c9d473"},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.12"},"toc":{"nav_menu":{},"number_sections":true,"sideBar":true,"skip_h1_title":false,"title_cell":"Table of Contents","title_sidebar":"Contents","toc_cell":false,"toc_position":{},"toc_section_display":true,"toc_window_display":false},"colab":{"provenance":[{"file_id":"1Bmp5vMvt0bfPhQE02XAD9z7QLpanD231","timestamp":1694015632597}]}},"nbformat":4,"nbformat_minor":5}